File: binarysearch.io/September 2020/stack_sequence.cpp
/*
Problem Link: https://binarysearch.io/problems/Stack-Sequence
Solution by Keshav Mishra
*/
#include "solution.hpp"
using namespace std;


class Solution {
    public:
    bool solve(vector<int>& pushes, vector<int>& pops) {
        int n1 = pushes.size();
        int n2 = pops.size();
        int i ,j;
        i = j = 0;
        vector<int> stack;
        while (i < n1 and j < n2){
            stack.push_back(pushes[i]);
            while(!stack.empty() && j < n2 && stack.back() == pops[j]){
                stack.pop_back();
                j++;
            } 
            i++;
        }
        while(!stack.empty() && j < n2 && stack.back() == pops[j]){
            stack.pop_back();
            j++;
        }
        return j == n2;
    }
};

File: binarysearch.io/stack_sequence.py
"""
Problem Link: https://binarysearch.io/problems/Stack-Sequence
Solution by Keshav Mishra
"""
class Solution:
    def solve(self, pushes, pops):
        # Write your code here
        n1 = len(pushes)
        n2 = len(pops)
        i = 0 
        j = 0 
        stack = []
        while i < n1 and j < n2:
            stack.append(pushes[i])
            while len(stack) and j < n2 and stack[-1] == pops[j]:
                j += 1 
                stack.pop()
            i += 1 
        while len(stack) and j < n2 and stack[-1] == pops[j]:
            j += 1 
            stack.pop()
        
        return j == n2 